File: 3098/main.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func main() {
	fmt.Println(sumOfPowers([]int{1,2,3,4}, 3))
}

func sumOfPowers(nums []int, k int) int {
	const mod = 1e9 + 7
	const inf = 0x3f3f3f3f
	n := len(nums)
	res := 0
	d := make([][]map[int]int, n)
	for i := range d {
		d[i] = make([]map[int]int, k + 1)
		for j := range d[i] {
			d[i][j] = make(map[int]int)
		}
	}
	sort.Ints(nums)
	for i := 0 ; i < n; i++ {
		d[i][1][inf] = 1
		for j := 0; j < i; j ++ {
			diff := int(math.Abs(float64(nums[i] - nums[j])))
			for p := 2; p <= k; p++ {
				for v, cnt := range d[j][p-1] {
					d[i][p][min(diff, v)] = (d[i][p][min(diff, v)] + cnt) % mod
				}
			}
		}
		for v, cnt := range d[i][k] {
			res = (res + v * cnt % mod) % mod
		}
	}
	return res
}

File: 3098/main.rs
use std::collections::HashMap;

struct Solution {}

impl Solution {
    pub fn sum_of_powers(nums: Vec<i32>, k: i32) -> i32 {
        const MOD : i64 = 1_000_000_007;
        const INF: i32 = 0x3f3f3f3f;
        let mut nums = nums.clone();
        let n = nums.len();
        let mut res = 0;
        let mut d: Vec<Vec<HashMap<i32, i64>>> = vec![vec![HashMap::new(); k as usize + 1]; n];
        nums.sort();
        for i in 0..n {
            d[i][1].insert(INF, 1);
            for j in 0..i {
                let diff = (nums[i] - nums[j]).abs();
                for p in 2..=k {
                    let mut updates: Vec<(i32, i64)> = Vec::new();
                    for (&v, &cnt) in &d[j][p as usize - 1] {
                        let key = diff.min(v);
                        updates.push((key, cnt));
                    }
                    for &(key, cnt) in &updates {
                        let entry = d[i][p as usize].entry(key).or_insert(0);
                        *entry = (*entry + cnt) % MOD;
                    }
                }
            }
            for (&v, &cnt) in &d[i][k as usize] {
                res = (res + v as i64 * cnt % MOD) % MOD;
            }
        }
        res as i32
    }
}

fn main() {
    println!("{}", Solution::sum_of_powers(vec![1,2,3,4], 3));
}